frontend/src/services/http/http.service.ts

```diff
@@ -1,18 +1,42 @@
 import { HttpError } from 'exceptions/exceptions';
 import { HttpHeader, HttpMethod } from 'common/enums/enums';
 import { HttpOptions } from 'common/types/types';
+import { PostInterceptor, PreInterceptor } from './interceptors/interceptor';
+import { refreshTokenInterceptor } from './interceptors/refresh-token-interceptor';
 
 class Http {
-  async load<T = unknown>(url: string, options: Partial<HttpOptions> = {}): Promise<T> {
+  private readonly refreshTokenHitStatusCode = 401;
+
+  async load<T = unknown>({
+    url,
+    options = {},
+    preInterceptors = [],
+    postInterceptors = [refreshTokenInterceptor],
+  }: {
+    url: string;
+    options?: Partial<HttpOptions>;
+    preInterceptors?: PreInterceptor[];
+    postInterceptors?: PostInterceptor[];
+  }): Promise<T> {
     const { method = HttpMethod.GET, payload = null } = options;
     const headers = this.getHeaders();
-
-    return fetch(url, {
+    let requestInit: RequestInit = {
       method,
       headers,
       body: payload,
-    })
-      .then(this.checkStatus)
+    };
+    for (const preInterceptor of preInterceptors) {
+      [url, requestInit] = await preInterceptor(url, options);
+    }
+
+    const makeRequest = (): Promise<Response> => fetch(url, requestInit);
+
+    let response = await makeRequest();
+    for (const postInterceptor of postInterceptors) {
+      response = await postInterceptor({ initialRequestFn: makeRequest, response });
+    }
```

Comment:
It seems for each request you refresh auth token pair. If not could you please give some info hot you attach access token in the header of each request?

Reply:
Yes, I forgot to add code to attach access tokens. Basically, now the flow is like the following:
- there are pre and post interceptors
- there is a preinterceptor for attaching access token header to request if access token exists
- also, postinterceptor for refreshing the token pair. it will only work if the initial response returned 401 code. and this code will be returned if the access token expired or wasn't attached (i.e. didn't exist)
- these interceptors are included by default, and there is no harm even if the endpoints don't require auth (because refreshing token pair will only happen if they return 401). they can, however be turned off, like in auth controller, where we don't want to refresh the tokens, because the endpoints return 401 on incorrect credentials.

Reply:
Please add pre-interceptor for attaching access token because in current implementation your logic work as I described in previous comment

Reply:
Please add pre-interceptor for attaching access token because in current implementation your logic work as I described in previous comment